lms/djangoapps/discussion/rest_api/api.py

```diff
@@ -1250,3 +1254,81 @@ def delete_comment(request, comment_id):
         comment_deleted.send(sender=None, user=request.user, post=cc_comment)
     else:
         raise PermissionDenied
+
+
+def _get_course_discussion_user_stats(
+    course_key: CourseKey,
+    order_by: UserOrdering = UserOrdering.BY_FLAGS,
+):
+    """
+    Get course user discussion stats.
+
+    Args:
+        course_key ():
+        order_by ():
+
+    Returns:
+
+    """
+    course_stats = get_course_user_stats(course_key)
+
+    def key_func(user):
+        key = (user["threads"], user["responses"], user["replies"])
+        if order_by == UserOrdering.BY_FLAGS:
+            key = (user["active_flags"], user["inactive_flags"]) + key
+        return key
+
+    return sorted(course_stats.values(), key=key_func, reverse=True)
```

Comment:
Since this is data about **all** users in a course, it could be quite large, but still I assume mostly under 100,000 entries. 
In which case the data should be in the 2-5MB range, and sorting should still be instantaneous. 

If this data is cached in memcache then that could be a fair bit of data across courses. 
Another option is to cache this in the database. We can keep it at 1 row per user per course, which would allow sorting and limiting this data directly from the database. 

In that case we'll need some mechanism to keep this in sync. We can probably schedule a task to run every few hours (12-24) to sync up this data. However, even than, running this once a day for the many many courses, each with a large number of users. The performance impact is something I'm concerned about and would want feedback on. 

Ideally we can update cs_comments_service to track this data so it's always up date and the query is a small one.

Reply:
That is one thing I was wondering. If it's better to do this sorting on edx-platform side where we can cache it better as well.

If we paginate and cache in cs_comments_service, then either we need a way to cache stuff on the ruby side (I'm not very familiar with that setup, I don't think there is much support there) or each time you change the page, the whole query is re-executed anyway.  It'll save on network transfer, but it's important to remember here that the whole data will still need to be transferred from mongodb to cs_comments_service anyway, I don't see a way to paginate that. 

On the other hand, on the edx-platform side, we can do one network request, cache that, and then pagination and sorting can reuse that data. 

So overall we save more network traffic in this way.

Reply:
I don't think a cache would be needed in ruby. This query should be optimized in mongodb itself, perhaps using a materialized view of the aggregation, refreshed by a cronjob (the mongodb query would need to be refactored to include that manual sum that's currently being done in ruby, but this is trivial).

Reply:
Currently, one of the goals is to avoid making major changes to the data structure in this repo. We can definitely optimise this a lot by caching user stats as a user interacts with discussions. That may just be the only performant solution, and we'll consider it if nothing else works.